### ISSUE TITLE
Identify credits for the next year of `gke.mybinder.org`

### ISSUE THREAD

**Issue by choldgraf** [jupyterhub/team-compass#463]
### Proposed change

Our annual allotment of credits for `gke.mybinder.org` runs out in late December (I believe, December 22nd 2021). We won't spend the credits down to 0 at that time, but they will expire on that date.

We need to identify where another round of funding for `gke.mybinder.org` will come from.

### Draft of two pager

[**See here for a draft two-pager to send to karan**](https://docs.google.com/document/d/1DvW8TYgEVWYvsgZKlr4JrmuhLQoYC-jTie0okgnIjp0/edit?usp=sharing)

### Action plan

We have 3 known options to power gke.mybinder.org:

1. GCP credits, if they extend them by another year
2. ~Jupyter Meets the Earth grant, if we can get approval by @fperez depending on whether they're in scope~ - this is likely not an option because it only runs on AWS
3. Pangeo cloud credits

here's the current plan:

- **On Wednesday, December 22nd** our credits on gke.mybinder.org run out.
- We wait until 6PM US/Pacific on Monday the 21st for a response from GCP and/or to confirm that we can use the JMTE funds, and don't take technical action until then.
- After then: If GCP gives us credits, then we direct the new billing account w/ credits to gke.mybinder.org, and we're golden.
- If not, but we have JTME approval, then we direct the JMTE billing account to gke.mybinder.org, then we're golden (for a month)
- If no JMTE approval, then we immediately switch into "deploy with Pangeo funding" mode and hope/plan for this to be completed by Wednesday the 22nd.
- If for some reason we don't think any of the above cases will work, then here's a blog post draft we should release, saying that capacity will drastically shrink [**Here is a draft blog post**](https://blog.jupyter.org/mybinder-org-capacity-temporarily-reducing-ce0390e0f0d4). We should then shut down `gke.mybinder.org` (unless somebody wants to foot the bill, I cannot pay for this deployment from my credit card again). This will reduce mybinder.org's capacity by about 75%, but I think that's just the reality that we face. Unless somebody has access to a large amount of Google Cloud credits that we can hook into gke.mybinder.org, I'm not sure what else we can do.

Regardless of all this, we need to release a blog post about the current situation, because it is clearly unsustainable (at least, for me it is unsustainable, and I assume for others as well)

### Tasks to complete

- [x] Put together a 2-pager that demonstrates Binder's impact
- [ ] Comments and feedback on the 2-pager (see link above)
- [ ] Wait for Karan's feedback

**Comment by choldgraf** [comment]
### Update: conversation with Karan and the GCP Research team

@consideRatio and I had a conversation with Karan from Google Cloud. He said that he was hopeful they'd be able to fund `gke.mybinder.org` for another round of cloud credits. In order to explore this, he'd need a 2-pager style document that demonstrated the impact of mybinder.org as well as the costs over time.

In particular, they care about things that demonstrate diverse and worldwide impact, like:

- Usage from communities outside of the US/Europe
- Total number of weekly sessions
- Number of github repositories served

I've updated the top comment with some next steps about putting together this 2-pager

**Comment by choldgraf** [comment]
I am going to try putting together a 2-pager ASAP that we can send to Karan, because 1.5 months is not that much time for us to get another round of funding. I would really appreciate any suggestions or help from others! Here are a few things that could be useful to help:

- Analyze some Binder data from the last year and come up with plots that demonstrate impact along the categories described above
- Come up with ideas for information we could put into a 2-pager that would demonstrate impact
- Help with writing or editing drafts

**Comment by minrk** [comment]
I'll work on gathering some analytics

**Comment by MridulS** [comment]
@minrk I did some work here https://gist.github.com/MridulS/5accc696311c4f381c05cb70922d362
<img width="912" alt="Screenshot 2021-11-17 at 12 53 30" src="https://user-images.githubusercontent.com/5363860/142195889-ff456e6a-de44-4533-aee5-b710005a5271.png">

**Comment by sgibson91** [comment]
I wonder if we can deploy this to the federation too, to make info gathering easier in the future? https://github.com/bitnik/binder-launches (Unfortunately, the link to the instance at GESIS no longer seems to be up)

**Comment by minrk** [comment]
Still working on analytics, since I've never really dug into matomo before, but here are monthly visits to GKE by continent:
![visualization-9](https://user-images.githubusercontent.com/151929/142262042-a9acf850-5f43-4e4c-a3a4-7babf6c94838.png)

and summing all non-NA-EU together shows it's about 50% NA, 30% EU, 20% rest-of-world:
![visualization-10](https://user-images.githubusercontent.com/151929/142263775-733d9213-a46c-4493-8227-a08bc78e655a.png)

https://gist.github.com/28e6c3aeb9e7a208e0986a67892e912d

**Comment by choldgraf** [comment]
Hey all - thanks for these very helpful graphs! I tried to reproduce some of them but I cannot figure out how to get the Matomo secret to access that data (here's an issue I opened as a result: https://github.com/jupyterhub/team-compass/issues/473#issuecomment-974558681).

Can anybody help me get access to the Matomo data so that we can include country information in this report?

**Comment by choldgraf** [comment]
### Update: draft is ready

Hey all - I took some of the plots here (some directly, some as inspiration) and put together the 2-pager at the link below:

https://docs.google.com/document/d/1DvW8TYgEVWYvsgZKlr4JrmuhLQoYC-jTie0okgnIjp0/edit?usp=sharing

I'd love feedback from folks if they think this looks OK. The goal of the 2-pager is to demonstrate the impact and usage of Binder, but doesn't need to go into a ton of detail. Also note, I couldn't figure out how to get Matomo data myself, so I just went with copy/pasting Google Analytics images, but happy (and prefer) to use Matomo data if somebody can help me get access to it.

**Comment by choldgraf** [comment]
### Update: sent to Karan for feedback

I know that this is a short turnaround, but we only have about a month before gke.mybinder.org runs out of credits, so I have sent the two-pager above to Karan for some feedback to see if we need to add anything to the 2-pager before he submits internally. I've cc'ed @minrk (as team lead) and @consideRatio (since he's been helping with the GCP Binder move lately) on the email. Will report back with relevant information.

**Comment by betatim** [comment]
Thanks a lot for putting the numbers together and adding words! I think it is good enough that we could send it already, so now we have a bit of time to make it even better.

I read the draft and left a few comments. Most of them are suggestions/nitpicks.

One thing I was wondering is if we can show/say something about an exciting/new area that is growing in terms of mybinder.org usage. The prime example that comes to my mind is things like executable books where we provide a crucial bit of infrastructure for courses/educational books from around the world that lets them do something that is otherwise super hard to do (executable sections in a text book). But I am not deeply enough into the executable books project/user base to know if there are a handful of neat projects we could point to. Not in detail but as a "this is a new area that is growing and super cool!"

**Comment by choldgraf** [comment]
Hey all - I have still not heard any specific response from Google, and so I want to start contingency planning for what to do if we do not get new credits in time. Here is what I propose:

# Timeline for running out of credits

- **On Wednesday, December 22nd** our credits on gke.mybinder.org run out.
- If we don't have word by tomorrow (Tuesday, December 21st), we should follow this plan:
  1. make a blog post on the Jupyter Blog telling users what to expect. [**Here is a draft blog post**](https://blog.jupyter.org/mybinder-org-capacity-temporarily-reducing-ce0390e0f0d4).
  2. On Wednesday the 22nd, shut down `gke.mybinder.org` (unless somebody wants to foot the bill, I cannot pay for this deployment from my credit card again)

This will reduce mybinder.org's capacity by about 75%, but I think that's just the reality that we face. Unless somebody has access to a large amount of Google Cloud credits that we can hook into gke.mybinder.org, I'm not sure what else we can do.

**Comment by betatim** [comment]
Sad times.

We will also need to find a new host for https://github.com/jupyterhub/mybinder.org-deploy/tree/master/images/federation-redirect and tweak its configuration so it will continue to work without GKE as the "prime". I think it makes sense to use the OVH deployment as the new prime site. I think these tasks need to happen to make the move:
* [ ] deploy federation proxy to new cluster
* [ ] double check the config
* [ ] change the IP address that mybinder.org points to

I think we can run two instances of the federation proxy in parallel without weird stuff happening. This means it shouldn't be a huge interruption to users.

Where/how should I give feedback on the blogpost draft?

Should we now tweet about the upcoming change? - If we do we give people (who read the tweet) only about 48h notice which isn't a lot. But hopefully there aren't too many people who rely on gke.mybinder.org explicitly or were planning big demos or some such. I think it would be a good idea to do so.

**Comment by choldgraf** [comment]
We are scrambling a bit to see if we can make up any extra funding from a different source. I also hope to have a more definitive answer from GCP by the end of day US/Pacific. There are two potential other funding sources we might be able to use in a stop-gap fashion.

- The Jupyter Meets the Earth grant (which we do not yet have explicit approval for, so need that first from @fperez if he thinks it's in-scope)
- The Pangeo cloud grant (which we do have approval to use from @rabernat, but which would require some technical complexity because that funding is parked at Columbia)

Either case it not a long-term solution, more like a 1-month stopgap to keep the lights on.

Here's my proposed plan:

- We wait until 6PM US/Pacific for a response from GCP and/or to confirm that we can use the JMTE funds, and don't take technical action until then.
- After then: If GCP gives us credits, then we direct the new billing account w/ credits to gke.mybinder.org, and we're golden.
- If not, but we have JTME approval, then we direct the JMTE billing account to gke.mybinder.org, then we're golden (for a month)
- If no JMTE approval, then we immediately switch into "deploy with Pangeo funding" mode and hope/plan for this to be completed by Wednesday the 22nd.

**Comment by betatim** [comment]
What does "deploy with Pangeo funding" mean? Switching billing accounts or deploying to a new cluster or something third?

**Comment by choldgraf** [comment]
@betatim yep, Pangeo has some grant funds parked at Columbia which are earmarked for a Binder deployment, and we can realistically say it is in-scope for that grant to pay for a short time of mybinder.org. However it'd require setting up a new project under the Columbia.edu cloud org, and re-deploying gke.mybinder.org there.

**Comment by betatim** [comment]
(sorry I edited my last comment above for a long time without clicking "save")

**Comment by betatim** [comment]
Has anyone asked the current members of the federation how much spare capacity we have there? Maybe we can increase our allocations there to make up for the lost capacity at GKE.

cc @MridulS for gesis, @sgibson91 for Turing (can you tag the right new person please?) and @mael-le-gal for OVH

**Comment by choldgraf** [comment]
re: action plan, my thinking was to wait until 6PM US/Pacific to gather information before making a decision/taking next steps, since most (all?) people that would be taking action are already on time zone where it is night time anyway, and this would allow us to spring into action in people's morning times if necessary.

re: other federation members, we've asked the Turing, but apparently they're close to running out of credits themselves :-/ not sure about GESIS or OVH.

For some quick numbers from [status.mybinder.org](https://mybinder.readthedocs.io/en/latest/about/status.html), it seems like we are talking about a pretty hefty increase in capacity to make up for it

![image](https://user-images.githubusercontent.com/1839645/146815825-6c8347a0-bc13-45fe-b140-138b61d55928.png)

**Comment by betatim** [comment]
Fine with me to wait till tomorrow morning to take action. If possible we should try and decide before then what it is we will do to give as many people as possible to 👍  or 👎. For example if we don't try and decide ~now/soon tomorrow morning in Asia and then Europe we need to either decide without considering US people or wait a long time for them to wake up.

I think the distribution of user pods across the clusters is biased by two things: the quota we set and the fact that until the quota is hit we send all launch requests for one repo to the same federation. This means who ever gets a popular repo (like the top ones from try.jupyter.org) will get a lot of traffic in one block. And if you don't get that you might be missing a big block.

Something else I don't remember us ever putting all too much thought into is the actual quotas. Should we rethink them? For example Gesis is a big cluster and we set a quota of 20. Typo? Decision? Possible to increase by a factor 5 or 10? I don't know so it would be cool to discuss that.

**Comment by choldgraf** [comment]
I agree with your assessment, though I think realistically I'm the only US-based person who is likely to be paying close attention here, and I don't understand the current cloud setup well enough to know what the right technical decisions are. I trust whatever the team decides if we get to that point. My focus is going to be on trying to get a definitive answer from GCP and the JMTE grant, since those would be our "least work" solutions if they are possible

**Comment by sgibson91** [comment]
@betatim Turing currently only has $800 😕 @callummole is following up

**Comment by manics** [comment]
Could we redirect the `jupyter/try` repos to JupyterLite https://discourse.jupyter.org/t/https-jupyter-org-try-so-slow-why/11288/7 ?

**Comment by choldgraf** [comment]
### Update: The JMTE grant likely won't work

I just heard from @consideRatio that the JMTE grant is only earmarked for AWS, not GCP, so this is likely not a reasonable short-term solution.

**Comment by choldgraf** [comment]
@manics sure that sounds great to me, provided the environment still works the same

**Comment by minrk** [comment]
> Could we redirect the jupyter/try repos to JupyterLite  

That would be a huge load off, and an interesting approach to take. Even if they're not the same, shifting that _traffic_ even as a temporary measure, might help, since it would greatly reduce the short-term cost of operating the main cluster. We can decide after things are stable if we want to keep it that way.

I should be around to work on this during the day in Europe, tomorrow, whatever the decision ends up being. If I have the necessary permissions!

**Comment by minrk** [comment]
The main truly GKE things (I think) are:

1. the events archive, and
2. the matomo managed SQL backend

If we lost those, even temporarily, moving to another federation member as the 'lead' would keep things from falling down, even if we had vastly diminished capacity. That shouldn't be too hard to change in the repo config.

Our event data 'going dark' isn't the best, but it's not the worst thing in the world, especially if it's temporary. Plus, the current time of year is also our lowest traffic, typically.

**Comment by manics** [comment]
@bollwyvl What would it take to replace some of https://jupyter.org/try with JupyterLite?

TLDR: The main mybinder.org member _may_ run out of credits on Wednesday! If it does  we want to reduce traffic as much as possible.

**Comment by minrk** [comment]
I opened https://github.com/jupyter/jupyter.github.io/issues/513 for shifting links to JupyterLite

**Comment by bollwyvl** [comment]
Chimed in over on the issue: happy to help with whatever we need.

**Comment by minrk** [comment]
I'm going to run https://github.com/jupyterhub/mybinder.org-deploy/pull/2012 to delete very-old images (before 2021-02-01) because our storage costs are currently $1600/mo and it takes a very long time to delete thousands of images from an image registry. So I want to have a lot of the super old stuff out of the way before the clock is really ticking and our only choice to stop bleeding money is to shut it all down. Plus, then we'll have a better sense of how long it will take to clear it all out.

**Comment by MridulS** [comment]
For the GESIS deployment, yes we can take much more than 20 pods (it's 80 now, we up it to 100 later if everything is stable). But the way the k8s cluster is currently designed/VM'd (one huge server + couple of small VMs, all the binder pods will be running in one node!), the cluster hits issues (specifically with creating and deploying containers)  when too many concurrent requests are sent GESIS way. 

I'll try to come up with something tomorrow EU morning (maybe add a couple of test servers to the cluster to increase the capacity, if possible).

**Comment by choldgraf** [comment]
Hey all - just wanted to give another update since I am signing off in a sec.

I still don't have a confirmation in credits from anybody in gcp. I do have a few leads and am having some more conversations tomorrow. I continue to have generally positive signals, but nothing definitive and no timeline.

Since I'll likely be asleep while folks are discussing this, I'm fine with whatever decision folks take to move forward. If we can get the gke deployment down to as low of a monthly rate as possible, I can foot the bill for a few weeks and try to reimburse myself from a grant somewhere. That might be a reasonable stopgap to keep us from having to move over all the matomo/archive/etc stuff. As long as we get the gke cost to less than, say 500 a month I am ok with that as a short term solution.

There are two other things we'd discussed that aren't encoded in the top comment:

1. It sounds like gesis can have a modest capacity boost. @MridulS maybe you can coordinate with folks about a reasonable limit to set?
2. It sounds like we can try directing people to jupyterlite for the top links on try.mybinder.org . I think we should try this tomorrow, as it would cut our cloud usage by about half. I am happy to spend some time tomorrow working on notebook documentation for those repositories, if any changes need to be made. Maybe that's something that @bollwyvl  or @jtpio could help with too?

Perhaps some combination of the above two things + finding a small pot of money somewhere will give us enough breathing room to move forward for a bit.

I'll check my email again before bedtime in case I hear back from anybody. If anybody else knows of an easy way to get quick cloud credits on gcp, please chime in - at this point even a few thousand would help us have a little time so we don't have to stress over the holidays (assuming we could just connect it to our project)

**Comment by minrk** [comment]
I'm also working on some leads for the small pot of funding that would let us keep everything as-is for the short term, and reducing costs to make that an easier goal.

FWIW, the docker registry api image-deletion was going to take at least a few days, but I found a simpler way to do it via the `gclould container images` API that seems to be going way faster.

**Comment by callummole** [comment]
Turing now has £5547.87 to help with the short term capacity. I can apply for more in the new year.

**Comment by minrk** [comment]
Terrific, thanks! Should we increase the turing quota? What do you think is a reasonable new capacity? 150 or 200? Do you know what the monthly cost of turing is, now?

**Comment by minrk** [comment]
Simula can cover up to a month of GKE if we don't come through with any last minute credits, so we don't need to do any drastic migrations. Just need to hook up a new billing account. 

For that, I just need to know if I have permission to do that, and exactly what time I should make the switch. Related: are the current credits on the project or the billing account (I.e. would we lose credits if I switched early?)

**Comment by callummole** [comment]
The the last couple of months the cost has been around $2k per month. So if the current capacity is 80 then perhaps we can could cover 200 for a month while we get more funds?

**Comment by sgibson91** [comment]
Would love for Turing in general to think about how we can increase that cluster. The 80 quota was an arbitrary decision I made when setting it up

**Comment by callummole** [comment]
@sgibson91 I'll investigate.

**Comment by minrk** [comment]
@callummole great! Feel free to make a PR like https://github.com/jupyterhub/mybinder.org-deploy/pull/2091 for turing when you come up with a reasonable number. It's okay to bump incrementally, too.

**Comment by choldgraf** [comment]
@minrk's plan sounds great to me!

Do we still wish to try making a switch to jupyterlite today? I am happy to spend time trying to get the right docs set up to make this happen if we think it'd be a big help

**Comment by minrk** [comment]
I think it's still worth doing. Main downside I see is that if there are issues, we don't want to be scrambling to respond to them over the holiday. Upside: lower traffic over the holiday!

I will be focusing on reducing the cost of keeping GKE as-is alive, without needing to squeeze _too_ hard.

As I see it right now, our costs break down _roughly_ 20% each:

- cpu
- ram
- pd-ssd
- cloud storage (almost all expired image cache, I believe)
- 'other' (SQL, non-ssd PDs, managed GKE fees, egress)

The first 3 are basically 'node costs' so if we reduce the number of nodes running, which is mostly proportional to the pods, we reduce all of those. Many of the 'other' fees would similarly scale down as well, but not all. The events archive itself takes up about as much space as 1-2 images, so not worth thinking much about.

Right now, I'm looking at whether our current 1TB SSD per node is useful, or if we should turn that down, and the perennial question of a more cost effective node flavor.

**Comment by minrk** [comment]
Our core nodes have 250G SSD boot disks, of which the most used is 8G, so that's a bit of a waste. But it pales in comparison to the user nodes, with have two SSDs each - 1TB main disk and 400GB disk just for dind builds. Our intermediate-age user node (gke-prod-user-202009-b9c03ca0-0hbq) has used 85% of the main disk and 50% of the dind disk, and it's only 4 days old. 

Note, these disks are _caches_ so they will always fill up eventually. Being full doesn't mean we need them as big as they are, it just means that when they fill up, they will start things like ImageGC to free space. A smaller disk means these things will run more often.

**Comment by minrk** [comment]
\*note: Local SSDs are fixed size, and billed as a separate SKU. Only ~$250/mo compared to $2k for the PD-SSDs.

Reducing the boot SSDs to 500GB per node would save ~$1k/mo. Increasing the node size would also save on storage, but we are already running near the limit of ~100 pods per node, so that wouldn't help.

**Comment by betatim** [comment]
Making the disks smaller sounds like a good plan. One thing I have at the back of my mind is that IO performance is linked to disk size. So maybe that was the reason for having such large disks (and we seem to end up with spare credits at the end of the year any way -> the time limit of the credits is a bigger factor that the amount). Worse performance is better than no performance though, so yay to smaller disks.

We could also ditch the "two disk" approach and use only the main disk to save even more money. I think OVH has been running in that mode for a while now. It needs a bit of a reconfiguration of the image GC to use an absolute size and not an inode based threshold.

**Comment by minrk** [comment]
The local SSD is a relatively small cost, so I'm not sure it's an optimization worth making right now. Since the same capacity on the PD SSD is 5x as expensive, merging the two probably doesn't make sense.

It _would_ be interesting if we could get the _host_ docker onto another local SSD and lose the PD-SSD altogether. That would would save _tons_ at a cost of fixed capacity per node. Not sure if that's possible, though.

**Comment by choldgraf** [comment]
I just got off the phone with the Google OSPO office. They are working to find stop-gap funding (maybe 6 weeks or so) in order to keep Binder running through January (though, if we can bring down the costs, we might be able to extend this a few months). That would buy us some time to work out a longer-term solution that is more sustainable for us (and for them) than this "every 1 year we frantically email people we know at Google" approach thus far. No promises from them, but I'm hopeful we'll work something out and will report back here as I learn more.

**Comment by minrk** [comment]
New quota increases from federation members have greatly reduced load on GKE prod. I've helped encourage scale-down a little with some cordoning. But I think between (ongoing) stale image deletions and load redistribution, we're looking at at least a few thousand dollars saved today on the monthly bill.

**Comment by minrk** [comment]
One thing we probably still need to do to run in a cost-conscious way is help scale-down with manual cordoning of low-occupancy user nodes. We had _7_ 42-day-old nodes, which is capacity for ~600 users at our lowest traffic times, I think? That's definitely more than we needed.

**Comment by betatim** [comment]
For the downscaling we should investigate the custom scheduler we use and if something has changed there. It used to work well :-/ An alternative we've discussed is using node preferences (similar to what we do for "sticky" build pods) where we work out the "least busy node" and then add a anti-preference for that node to a pod when creating it.

**Comment by arnim** [comment]
If absolutely necessary and of help, GESIS would also be able to contribute ~$5k via Linode.

**Comment by betatim** [comment]
Somewhat off-topic but also not: does anyone know why https://jupyter.org/try links to https://mybinder.org/v2/gh/jupyterlab/jupyterlab-demo/HEAD?urlpath=lab/tree/demo which currently doesn't build? I thought this must have been a recent commit that broke it but turns out the last commit was in mid October. Maybe no one will notice/complain if we don't provide a seamless transition/switch it over to something a bit different (jupyterlite) given that it seems to have been broken for a while now?

**Comment by choldgraf** [comment]
@betatim nope I don't know who controls that repo or that link, but definitely agree that this is another reason to just use JupyterLite. Here's the issue @minrk brought up to discuss that: https://github.com/jupyter/jupyter.github.io/issues/513

To try and prep for that, I just made a PR to the JupyterLite docs to add a more "introductory" notebook for their links: https://github.com/jupyterlite/jupyterlite/pull/432

**Comment by minrk** [comment]
Looks like we got some interim credits from Google just in time, so we have a little slack while we work on the more permanent solution.

**Comment by minrk** [comment]
> which currently doesn't build?

I think a dependency must have updated out from under it. When I tested, it ran fine [on GKE](https://gke.mybinder.org/v2/gh/jupyterlab/jupyterlab-demo/HEAD?urlpath=lab/tree/demo) and didn't need a build. This is probably in our top 2 most popular images, so I think folks would notice. It must be getting assigned to turing more often with the recent changes, where it wasn't in the cache already.

pyyaml 6.0 recently dropped support for the long-deprecated, but still widely used `yaml.load(str)`, which is the source of the failure. Fixed in https://github.com/jupyterlab/jupyterlab-demo/pull/113

**Comment by choldgraf** [comment]
As @minrk - we just got $10,000 of Google Cloud credits deposited into the same GCP billing account, and they expire in 6 months. This means that we don't need to make any chances and the service will keep running.

I'd like to write a short blog post about this episode to give transparency to our user community about what happened and what we're doing to try and improve things in the future. I imagine something like:

1. Brief description of what happened
2. Cost reduction efforts (e.g. image cleaning etc)
3. Load balancing efforts (e.g., other federation members stepping up to carry more load)
4. Stop-gap credits from Google
5. Next step is exploring ways that we can further make progress on each of these items

Does anybody object to that plan? I'll try to get a draft ASAP while the experience is still fresh.

**Comment by betatim** [comment]
Hadn't thought about dependencies changing :-/

I think a blog post is a good idea. I would lead with (5) though instead of a chronological/experience report order. My reasoning is that (5) is the most important thing out of all this for the reader and what we'd like the reader to help us with. This in turn makes me wonder what we are looking for and if we can express that in a couple of sentences. Some properties that I think we want: (0) someone who enjoys fundraising (1) multi year (2) GKE and (3) <$10000 per month for sure, maybe $5000 per month in credits. (4) a call to action "contact us via this thing" if you can help with funding or the effort to secure funding.

Other things we could be doing: on board more federation members, increase the capacity at existing federation members. But I would put them in a separate blog post or further down in this post to focus on the above give points as the one thing people remember.

I would lead with our ask/next steps because no one reads stuff on the internet and even those who read, should get the most important point first.

---

Before writing I think we should sharpen what we are asking for and how people can reach us so that we have a very concrete idea for both. This will allow us to write a clear article with a concrete call to action. And means in the team we have alignment on it.

**Comment by manics** [comment]
Following on from @betatim's last point, also work out how much time (if any) we can devote to cases where compute can be contributed but not people (thinking of https://github.com/jupyterhub/mybinder.org-deploy/issues/1772)

**Comment by betatim** [comment]
I think for things like https://github.com/jupyterhub/mybinder.org-deploy/issues/1772 we will not find out if they are a net good/bad without trying it. But someone has to have time and drive to keep moving it forward.

**Comment by choldgraf** [comment]
Regarding a blog post, I was thinking of just a minimal "what happened, what we did to resolve it, and thank you Google" post rather than a more future-looking post. My reasoning is that I worry raising the bar too much will increase the likelihood that no post will happen at all.

I definitely agree it's a good idea to come up with an action plan, call to action, etc but do we have bandwidth to do this? Is anybody willing to champion this?

I do think the Google team is interested in meeting further in January to find a more sustainable solution, which is where I'm going to put my cycles if I have them.

**Comment by betatim** [comment]
> I do think the Google team is interested in meeting further in January to find a more sustainable solution, which is where I'm going to put my cycles if I have them.

Is there a way to help with that? More generally I am wondering what the next steps are here.

**Comment by choldgraf** [comment]
Thanks for following up @betatim - a few updates:

### Blog post / CTA
Yesterday I put together a short draft that tried to incorporate some of the ideas shared above. I added a section for "what we need / what you can do" but I suspect it'll need a bit of iteration:

https://docs.google.com/document/d/1A2TDXlQ1ap1dM7ek2gRRfSL9O6xudgNgOQqNW3LUwPo/edit?usp=sharing

If folks are interested in having a dedicated brainstorm to think about sustainable pathways forward, I'd be happy to so.

### Google credits

I got in touch with Karan yesterday to check up on the status of the credit request we had originally put in. He said they'd like to meet on Thursday to discuss more sustainable pathways forward. He put a meeting on my calendar for 9:30AM US/Pacific time. I don't think we want to overwhelm them with a ton of people, but if anybody is interested in joining this conversation as well I'd love to have you there. Just let me know! Either way, I will report back what we discuss after that conversation.

**Comment by choldgraf** [comment]
### Update from conversation with Google

Just had a quick meeting with a few folks from Google Research. Here are the notes:

https://docs.google.com/document/d/1W5q3WLeT_sviLrW0zhpmo5DfPww8sA0w7tlBCeeAa6o/edit?usp=sharing

`tl;dr`: we have roughly two options that we can explore in parallel:

1. Kick-off a credit request that is similar to the one they approved with fast turnaround in December. This could get mybinder.org credits for another year or so.
2. Explore a more formal "in-kind sponsorship", which would help formalize the relationship between Binder and Google more clearly, and make it easier / less work to renew this collaboration each year rather than blindly reaching out to folks inside of Google

They're going to start the process for 1 right now, and for 2 we'll need to do two things:

1. Engage in some kind of conversation with them in the next few months to agree on what an in-kind sponsorship would look like
2. Decide **for ourselves** what it means to "support the Binder project". If we could provide some structure for them (e.g., a "sponsorships page" with clear criteria for different sponsorship levels, and a list of organizations that support the project at each level), then this would make things easier for them and others to support us. They can give us feedback if we want to work on this kind of thing.

**Comment by choldgraf** [comment]
We've had a few conversations here and in the Matrix channel about sustainability opportunities to explore. Rather than ballooning this issue into a long thread about sustainability, I decided to update the top comment of #430 so that it captures a few of the ideas we've discussed for longer-term sustainability efforts.

Are folks OK taking the long-term sustainability conversation there, so we can focus this thread around extending our credit runway with Google in the short-term?

**Comment by choldgraf** [comment]
I'm going to close this one, as we have a resolution in https://github.com/jupyterhub/mybinder.org-deploy/issues/2138 and we've also got a longer-term plan for credits in these two issues:

- #508 
- https://github.com/jupyter/governance/issues/125

**Comment by minrk** [comment]
Super useful, thanks for chasing that down, @choldgraf!
